internal/engine/k8swatch/pod_watch.go

```diff
@@ -37,79 +65,196 @@ type PodWatcher struct {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	nsWatch, ok := w.watchedNamespaces[ns]
+	if !ok {
+		return false
 	}
+	return nsWatch.watchers[key]
 }
 
-func (w *PodWatcher) OnChange(ctx context.Context, st store.RStore, _ store.ChangeSummary) {
-	taskList := w.diff(ctx, st)
+// ExtraSelectors returns the extra selectors for a given KubernetesDiscovery object.
+//
+// This is intended for use in tests.
+func (w *PodWatcher) ExtraSelectors(key types.NamespacedName) []labels.Selector {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	var ret []labels.Selector
+	for _, s := range w.extraSelectors[key] {
+		ret = append(ret, s.DeepCopySelector())
+	}
+	return ret
+}
 
+// Claimant returns the key that has uniquely claimed a particular K8s entity UID to prevent
+// duplicate events.
+//
+// This is intended for use in tests.
+func (w *PodWatcher) Claimant(uid types.UID) types.NamespacedName {
 	w.mu.Lock()
 	defer w.mu.Unlock()
+	return w.claimedUIDs[uid]
+}
+
+// reconcile manages namespace watches for the modified KubernetesDiscovery object.
+func (w *PodWatcher) reconcile(ctx context.Context, st store.RStore, key types.NamespacedName) {
+	state := st.RLockState()
+	kd := state.KubernetesDiscoveries[key]
+	st.RUnlockState()
+
+	if kd == nil {
+		// this object doesn't exist in store (i.e. it was deleted) so clean up
+		// any watches + other metadata
+		delete(w.kubernetesDiscoveryManifest, key)
+		delete(w.extraSelectors, key)
+		for _, watcher := range w.watchedNamespaces {
+			// remove this key from the list of watchers; if it was the last watcher, the
+			// namespace watch will be cleaned up afterwards
+			delete(watcher.watchers, key)
+		}
+		for uid, k := range w.claimedUIDs {
+			if k == key {
+				delete(w.claimedUIDs, uid)
+			}
+		}
+		return
+	}
 
-	w.extraSelectors = taskList.extraSelectors
+	if manifestName, ok := kd.Annotations[v1alpha1.AnnotationManifest]; ok && manifestName != "" {
+		// PodChangeAction is currently tied to manifest so keep a mapping of API objects to manifests
+		// (this will be removed shortly once PodChangeAction -> KubernetesDiscoveryUpdateStatusAction)
+		w.kubernetesDiscoveryManifest[key] = model.ManifestName(manifestName)
+	}
 
-	for _, teardown := range taskList.teardownNamespaces {
-		watcher, ok := w.watcherKnownState.namespaceWatches[teardown]
-		if ok {
-			watcher.cancel()
+	w.extraSelectors[key] = nil
+	for _, labelValues := range kd.Spec.ExtraSelectors {
+		ls := selectorFromLabels(labelValues)
+		if !ls.Empty() {
+			w.extraSelectors[key] = append(w.extraSelectors[key], ls)
 		}
-		delete(w.watcherKnownState.namespaceWatches, teardown)
 	}
 
-	for _, setup := range taskList.setupNamespaces {
-		w.setupWatch(ctx, st, setup)
+	seenUIDs := make(map[types.UID]bool)
+	seenNamespaces := make(map[k8s.Namespace]bool)
+	for _, toWatch := range kd.Spec.Watches {
+		namespace := k8s.Namespace(toWatch.Namespace)
+		if !seenNamespaces[namespace] {
```

Comment:
this check is probably redundant because it's handled internally

Reply:
Cleaned this up and made the UID stuff more consistent as well, so there's `setupNamespaceWatch` and `setupUIDWatch` which are both idempotent, so it just calls them both indiscriminately - ebf7462ef